src/cdp/cdp.zig

```diff
@@ -0,0 +1,316 @@
+// Copyright (C) 2023-2024  Lightpanda (Selecy SAS)
+//
+// Francis Bouvier <francis@lightpanda.io>
+// Pierre Tachoire <pierre@lightpanda.io>
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+const std = @import("std");
+
+const server = @import("../server.zig");
+const Ctx = server.Cmd;
+
+const browser = @import("browser.zig").browser;
+const target = @import("target.zig").target;
+const page = @import("page.zig").page;
+const log = @import("log.zig").log;
+const runtime = @import("runtime.zig").runtime;
+const network = @import("network.zig").network;
+const emulation = @import("emulation.zig").emulation;
+const fetch = @import("fetch.zig").fetch;
+const performance = @import("performance.zig").performance;
+
+pub const Error = error{
+    UnknonwDomain,
+    UnknownMethod,
+    NoResponse,
+};
+
+pub fn isCdpError(err: anyerror) ?Error {
+    // see https://github.com/ziglang/zig/issues/2473
+    const errors = @typeInfo(Error).ErrorSet.?;
+    inline for (errors) |e| {
+        if (std.mem.eql(u8, e.name, @errorName(err))) {
+            return @errorCast(err);
+        }
+    }
+    return null;
+}
+
+const Domains = enum {
+    Browser,
+    Target,
+    Page,
+    Log,
+    Runtime,
+    Network,
+    Emulation,
+    Fetch,
+    Performance,
+};
+
+// The caller is responsible for calling `free` on the returned slice.
+pub fn do(
+    alloc: std.mem.Allocator,
+    s: []const u8,
+    ctx: *Ctx,
+) ![]const u8 {
+
+    // JSON scanner
+    var scanner = std.json.Scanner.initCompleteInput(alloc, s);
+    defer scanner.deinit();
+
+    std.debug.assert(try scanner.next() == .object_begin);
+
+    // handle 2 possible orders:
+    // - id, method <...>
+    // - method, id <...>
+    var method_key = (try scanner.next()).string;
+    var method_token: std.json.Token = undefined;
+    var id: ?u16 = null;
+    // check swap order
+    if (std.mem.eql(u8, method_key, "id")) {
+        id = try getId(&scanner, method_key);
+        method_key = (try scanner.next()).string;
+        method_token = try scanner.next();
+    } else {
+        method_token = try scanner.next();
+    }
+    try checkKey(method_key, "method");
+
+    // retrieve method
+    const method_name = method_token.string;
+    std.log.debug("cmd: method {s}, id {any}", .{ method_name, id });
```

Comment:
Yes, the log will display null if id it's not set at this stage.
Anyway I will reduce the logs when everything is working well so this might be removed.